cranelift/codegen/src/isa/aarch64/lower_inst.rs

```diff
@@ -798,13 +798,48 @@ pub(crate) fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
         }
 
         Opcode::Ishl | Opcode::Ushr | Opcode::Sshr => {
+            fn lower_shift_amt<C: LowerCtx<I = Inst>>(
+                ctx: &mut C,
+                amt_input: InsnInput,
+                dst_ty: Type,
+            ) -> ResultRegImmShift {
+                let amt_ty = ctx.input_ty(amt_input.insn, amt_input.input);
+
+                match (dst_ty, amt_ty) {
+                    // When shifting for amounts larger than the size of the type, the CLIF shift
+                    // instructions implement a "wrapping" behaviour, such that an i8 << 8 is
+                    // equivalent to i8 << 0
+                    //
+                    // On i32 and i64 types this matches what the aarch64 spec does, but on smaller
+                    // types (i16, i8) we need to do this manually, so we wrap the shift amount
+                    // with an AND instruction
+                    (I16 | I8, _) => {
+                        // We can ignore the top half of the shift amount register if the type is I128
+                        let amt_reg = put_input_in_regs(ctx, amt_input).regs()[0];
+                        let mask = (ty_bits(dst_ty) - 1) as u64;
+                        let tmp = ctx.alloc_tmp(I32).only_reg().unwrap();
```

Comment:
Another potential optimization here is to use the shift destination register instead of a temporary one.